chromium_src/components/history/core/browser/sync/typed_url_sync_bridge.cc

```diff
@@ -0,0 +1,49 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_sync/features.h"
+
+namespace {
+
+bool IsSendAllHistoryEnabled() {
+  return base::FeatureList::IsEnabled(
+      brave_sync::features::kBraveSyncSendAllHistory);
+}
+
+}  // namespace
+
+#define BRAVE_RETURN_TRUE_IF_SEND_ALL_HISTORY_ENABLED \
+  if (IsSendAllHistoryEnabled()) {                    \
+    return true;                                      \
+  }
+
+#define BRAVE_HAS_TYPED_URL_RETURN_TRUE_IF_SEND_ALL_HISTORY_ENABLED \
+  if (IsSendAllHistoryEnabled()) {                                  \
+    /* When there are no visits, we must return false. Otherwise */ \
+    /* TypedURLSyncBridge will try to send data to sync .*/         \
+    return !visits.empty();                                         \
+  }
+
+#define BRAVE_TYPED_URL_SYNC_BRIDGE_SHOULD_SYNC_VISIT_RETURN_TRUE_IF_SEND_ALL_HISTORY_ENABLED \
+  BRAVE_RETURN_TRUE_IF_SEND_ALL_HISTORY_ENABLED
+
+// At TypedURLSyncBridge::WriteToTypedUrlSpecifics when syncing all
+// history, don't ignore reload transitions
+#define BRAVE_WRITE_TO_TYPED_URL_SPECIFICS_DONT_IGNORE_RELOADS \
+  if (!IsSendAllHistoryEnabled())
+
+// TypedURLSyncBridge::ApplySyncChanges is invoked when data arrives from the
+// server into local storage. Therefore it is possible that other device in
+// chain will have kBraveSyncSendAllHistory feature enabled, but the current
+// device won't. So skip DCHECK below without respect to the feature.
+#define BRAVE_TYPED_URL_SYNC_BRIDGE_APPLY_SYNC_CHANGES_SKIP_DCHECK if (false)
```

Comment:
With `kBraveSyncSendAllHistory` enabled remote devices, they will send other types of records in the same `TypedUrlSpecifics` so I don't understand why we would fail this `DCHECK(specifics.visits_size());`.
Can you elaborate more on that?

Reply:
@darkdh 

You right, this is wrong. I can't see this case anymore, whatever I tried to do. Thank you for pointing this, I have removed this unnecessary patch at https://github.com/brave/brave-core/pull/16705/commits/ddd1ac8b1e4c602fd53a9e610c5771f386020d2e.

Here are the STR for how it was before.

The STR were:
1. disable BRAVE_TYPED_URL_SYNC_BRIDGE_APPLY_SYNC_CHANGES_SKIP_DCHECK macro and build
2. Device2 has `brave-sync-send-all-history` flag turned on, device1 has this flag default (disabled).
3. On device2 open new tab with brave.com website
4. Join devices into the sync chain, enable sync of history.
5. On device2 browser, switch to tab with brave.com and close device2 browser
6. Open device2 browser.

At this step I used to get `specifics` with `specifics.visits_size()==0`. Now I can't see it:
```
[126661:126678:0130/172225.284174:VERBOSE1:model_type_worker.cc(648)] Encryption Keys: Delivering 0 applicable updates.
[126661:126679:0130/172225.284390:ERROR:typed_url_sync_bridge.cc(239)] [BraveSync] ApplySyncChanges 000
[126661:126679:0130/172225.284434:ERROR:typed_url_sync_bridge.cc(273)] [BraveSync] ApplySyncChanges 001 url=https://brave.com/
[126661:126679:0130/172225.284448:ERROR:typed_url_sync_bridge.cc(274)] [BraveSync] ApplySyncChanges 002 specifics.visits_size()=1
[126661:126679:0130/172225.284463:ERROR:typed_url_sync_bridge.cc(276)] [BraveSync] ApplySyncChanges 003 specifics.visit_transitions(0)=PAGE_TRANSITION_RELOAD
```

It looks I used to see it before I had introduced `BRAVE_WRITE_TO_TYPED_URL_SPECIFICS_DONT_IGNORE_RELOADS` patch. So device 2 indeed could form and send empty visits, because reloads were ignored.

In anyway, now `BRAVE_TYPED_URL_SYNC_BRIDGE_APPLY_SYNC_CHANGES_SKIP_DCHECK` is pointless.